lib/hexpm_web/session.ex

```diff
@@ -1,65 +1,212 @@
+
+  def to_id(session, token) when byte_size(token) == 64 do
+    "#{session.uuid}#{Base.url_encode64(token)}"
+  end
+
+  def to_id(_, _), do: {:error, :invalid_session_token}
+
+  # Private API
+
+  defp get_active_session(id) do
+    case Repo.one(Session.get_active(id)) do
+      nil ->
+        {:error, :no_session}
+
+      session ->
+        {:ok, session}
+    end
+  end
+
+  defp clear_and_drop(conn) do
+    conn
+    |> clear_session()
+    |> configure_session(drop: true)
+  end
+
+  defp expire(nil), do: {:error, :no_session}
+
+  defp expire(%Session{} = session) do
+    session
+    |> Session.expire()
+    |> Repo.delete!()
+  end
+
+  defp hash_token(token), do: :crypto.hash(:sha256, token)
+
+  defp parse_session_str(session_str) do
+    with {:ok, {id, base64}} <- parse_id(session_str),
+         {:ok, token} <- decode_token(base64) do
+      {:ok, {id, token}}
     end
+  end
+
+  defp parse_id(<<uuid::binary-size(36), base64::binary-size(88)>>) do
+    case Ecto.UUID.dump(uuid) do
+      {:ok, _raw} ->
+        {:ok, {uuid, base64}}
 
-    build_cookie(id, token)
+      _ ->
+        {:error, :invalid_session_id}
+    end
   end
 
-  def delete(_conn, {id, _token}, _opts) do
-    if Repo.write_mode?() do
-      Repo.delete_all(Session.by_id(id))
+  defp parse_id(_), do: {:error, :invalid_session_id}
+
+  defp decode_token(base64) do
+    case Base.url_decode64(base64) do
+      {:ok, _token} = return ->
+        return
+
+      _ ->
+        {:error, :invalid_session_token}
     end
+  end
+
+  defp maybe_expire(session) do
+    now = DateTime.utc_now()
+
+    case DateTime.compare(now, session.expires_at) do
+      :lt ->
+        {:ok, session}
 
-    :ok
+      :gt ->
+        expire(session)
+        {:error, :expired}
+    end
+  end
+
+  defp validate_token(%Session{} = session, token) when byte_size(token) == 64 do
+    case Plug.Crypto.secure_compare(hash_token(token), session.token_hash) do
+      true ->
+        {:ok, session}
+
+      false ->
+        {:error, :invalid_token}
+    end
+  end
+
+  defp validate_token(_, _), do: {:error, :invalid_token}
+
+  defp setup_session(conn, id) do
+    case get(id) do
+      {:error, :no_session} ->
+        not_found(conn)
+
+      {:error, :invalid_session_id} ->
+        not_found(conn)
+
+      {:error, :invalid_token} ->
+        not_found(conn)
+
+      {:error, :expired} ->
+        expired(conn)
+
+      {:ok, session} ->
+        conn
+        |> assign(:current_user, session.user)
+        |> assign(:current_organization, session.user.organization)
+        |> configure_session(renew: true)
+    end
   end
 
-  defp build_cookie(session) do
-    build_cookie(session.id, session.token)
+  defp not_found(conn) do
+    conn
+    |> configure_session(renew: true)
+    |> assign(:current_user, nil)
+    |> assign(:current_organization, nil)
   end
 
-  defp build_cookie(id, token) do
-    "#{id}++#{Base.url_encode64(token)}"
+  defp expired(conn) do
```

Comment:
@ericmj Took out the put_flash and redirect here, but we may still want to do that for this case. Just let me know.